pkg/expansion/system.go

```diff
@@ -207,8 +220,13 @@ func (s *System) expand(base *mutationtypes.Mutable) ([]*Resultant, error) {
 	return resultants, nil
 }
 
-func expandResource(obj *unstructured.Unstructured, ns *corev1.Namespace, template *expansionunversioned.ExpansionTemplate) (*unstructured.Unstructured, error) {
-	if ns == nil {
+func expandResource(obj *unstructured.Unstructured, ns *corev1.Namespace, template *expansionunversioned.ExpansionTemplate, opts ...ExpandOption) (*unstructured.Unstructured, error) {
+	options := &ExpandOptions{}
+	for _, opt := range opts {
+		opt(options)
+	}
+
+	if !options.RelaxNamespaceNullabilityCheck && ns == nil {
```

Comment:
Why is namespace required for expansion? What about cluster-scoped resources?

Reply:
I'm going to wait on Davis to chime in here too before making changes but my current understanding is that we need to check namespace nullability if a constraint or mutator specifies a namespace selector. We really check the namespace so we can set it below on the expanded resource. So if we don't have a non nil namespace and we don't error out on that, we can create expanded resources without the namespace that constraints or mutators would look for, which can lead to under enforcement.

I'm going to defer to Davis for the cluster-scoped scenario.

Reply:
Cluster-scoped objects are identified by having empty-string for the namespace:

https://github.com/open-policy-agent/gatekeeper/blob/b7230e0bdc3479b867acd2fd0d7192b1182c59d4/pkg/mutation/match/match.go#L82-L85

Constraints have no effect on expansion.

If mutators have match criteria that expects a namespace and that namespace is missing, it will raise an error when the mutation is applied.

https://github.com/open-policy-agent/gatekeeper/blob/b7230e0bdc3479b867acd2fd0d7192b1182c59d4/pkg/mutation/match/match.go#L96-L98

Functionally, it is equivalent, and simpler, to rely on match criteria to raise an error if their requirements are not satisfied (no namespace-based selection, no error).

This means:

* drop validation of namespace presence in expansion templates
* `metadata.namespace` of the expanded object inherits its value (or lack thereof) from `metadata.namespace` of the original object, rather than `metadata.name` of a namespace object passed in to the expansion system.
   * relying on the namespace object being non-nil is broken for cluster-scoped objects
   * inheriting from `metadata.namespace` preserves the assumptions behind the if-namespace-name-then-namespace-object-must-be-non-nil check in the matcher code.

Reply:
I agree with Max, it would be easier to just [relax the nil check](https://github.com/open-policy-agent/gatekeeper/blob/master/pkg/expansion/system.go#L211).  Some of the downstream code might have sets stuff based on namespace will have to be put in conditionals.

Originally,  I confused myself in thinking ExpansionTemplates have `Match`, but they don't -- they  have `ApplyTo`. Unfortunately, I think I passed this confusion onto Alex :( Very sorry!

Reply:
hey @davis-haba @maxsmythe finally got around to updating this PR. PTAL at this comment and not the code.

At present, I followed Max's advice to drop the namespace nullability check altogether.

It seems that in not setting the namespace for a generated resource, the matcher system will think the resource is "cluster scoped". This effectively introduces a bug, where namespaced generated resources are not matched properly.

I'm going to have a look at how we can modify some of the logic in `match.go` to accommodate resultants that are namespaced BUT don't have a namespace set.

Reply:
Shouldn't generated resources inherit their namespace from the parent resource?

Reply:
As of now, we can get a message like

```
unable to match constraints: error matching the requested object: failed to run Match criteria: namespace selector for namespace-scoped object but missing Namespace
```

in a shift left scenario.